src/escape.rs

```diff
@@ -131,6 +155,116 @@ where
         }
         i += 1;
     }
-    w.write_all(&bytes[mark..])?;
+    w.write_all(&bytes[mark..])
+}
+
+/// Calls the given function with the index of every byte that is in
+/// [", &, <, >] and for no other byte.
+#[cfg(all(target_arch = "x86_64", feature="simd"))]
+fn scan_simd<F>(bytes: &[u8], mut offset: usize, mut callback: F) -> io::Result<()>
+    where F: FnMut(usize) -> io::Result<()>
+{
+    if bytes.is_empty() {
+        // we need this special case for when we are passed an empty subslice
+        // that is exactly on a page boundary. we would try to read on a page
+        // that is potentially not ours and crash
+        return Ok(());
+    }
+
+    let lower_vec = unsafe { _mm_set_epi8( 0, 62, 0, 60, 0, 0, 0, 0, 0, 38, 0, 0, 0, 34, 0, 127 ) };
+    let upperbound = bytes.len().saturating_sub(16);
+
+    while offset < upperbound {
+        let mut mask = unsafe {
+            let raw_ptr = transmute(bytes.as_ptr().offset(offset as isize));
+            let v = _mm_loadu_si128(raw_ptr);
+            let expected = _mm_shuffle_epi8(lower_vec, v);
+            let matches = _mm_cmpeq_epi8(expected, v);
+            _mm_movemask_epi8(matches)
+        };
+
+        while mask != 0 {
+            let ix = mask.trailing_zeros();
+            callback(offset + ix as usize)?;
+            mask ^= mask & -mask;
+        }
+
+        offset += 16;
+    }
+
+    // final iteration. there's two scenario's:
```

Comment:
This is a deep question, whether it's defined behavior to read from the same page. My personal feeling is that it's never ok to read a byte past the slice.

Also, I'm pretty sure this is not handling the empty case.

Here's the strategy I would employ:

If len < 16, use scalar fallback. There is basically no safe way to read SIMD here.

Process len / 16 chunks. Then if len % 16 != 0, read last chunk starting at len - 16, compute mask then shift right len % 16.

Reply:
Right. I don't yet have hard evidence for this, but I've read `libc`'s `strlen` basically does this as well. They kind of have to use this technique as C strings are zero terminated and the length is not known in advance (that would make the function kind of trivial :p). It supposedly reads past the end of the buffer, but never past a 16 byte bound.

I can dig into this further tho. I think I have seen it in other libraries as well. `jetscii` comes to mind.

Reply:
I can see libc needing it, but in our case we have the length. This is an area where I'd really love input from an expert. @burntsushi ?

Reply:
I ran into a similar decision point in memchr, and my conclusion was "absent compelling evidence to the contrary, this is UB, so don't do it." I ended up [falling back to a simple scalar loop](https://github.com/BurntSushi/rust-memchr/blob/d70ea8a1dedd89be028e11f09c99423d4393e26f/src/x86/sse2.rs#L115-L123) when the input was too small to fit into a vector. Note, though, that the scalar loop only kicks in when the overall input is too small, and _not_ when the number of remaining bytes to look at is too small. Instead, we just accept that the last check might overlap with a previous check, but otherwise ensure that we are correctly reading in a full vector's worth of data: https://github.com/BurntSushi/rust-memchr/blob/d70ea8a1dedd89be028e11f09c99423d4393e26f/src/x86/sse2.rs#L182 --- I haven't looked at this code too carefully, but I imagine you can do the same by shifting some things around.

In my research, I came across this relevant SO answer: https://stackoverflow.com/questions/37800739/is-it-safe-to-read-past-the-end-of-a-buffer-within-the-same-page-on-x86-and-x64 However, with phrases like "Reading outside an object is of course Undefined Behaviour in C, but it's well-defined for C-targeting-x86," it feels like the accepted answer is on dubious ground.

From what I can tell, one way this might be safe is if you dropped down into Assembly and did reads past the end of the buffer there. But up in Rust or C land, I feel like you're on much shakier footing.

With respect to jetscii, I had thought [PCMPESTRI](https://www.felixcloutier.com/x86/pcmpestri) had a special exception for this, but I don't remember where I saw it. Personally, I feel like if you at any point have a pointer to some data that is out of bounds in a Rust program, you've probably entered UB territory.

I'd like to stress that I do not actually consider myself an expert in this, and that I feel like a true canonical answer would probably need to come from someone who is involved in rustc/lllvm development with respect to these special vendor intrinsics. (I imagine that's a fairly small number of people.) In general, my stance is a reflection of 1) no conclusive evidence that this isn't UB and 2) generally being conservative.

Reply:
Thanks so much for your detailed considerations Andrew! All the references are very helpful too.

Just a single note regarding the below.

> Personally, I feel like if you at any point have a pointer to some data that is out of bounds in a Rust program, you've probably entered UB territory.

I'm unsure if you meant this, but in this case all pointers should point to a safe range. It's just that a 16 byte read from that pointer may read from outside that range.

Reply:
Ah yes, right, that's what I meant. :-)